EnMasseProject/enmasse
# Add operator metrics

### Type of change
- Enhancement / new feature

### Description
Relevant issue - https://github.com/EnMasseProject/enmasse/issues/3686
Adding operator metrics to the enmasse operator that are provided by default with the newer versions of the operator SDK. A service and serviceMonitor are created for some prometheus instance to scrape the metrics.

### Verification Steps
1. Install the operator
2. Replace operator image with `quay.io/davidffrench/controller-manager:add-operator-metrics`
3. verify a service and service monitor are created. Both with the label `monitoring-key=middleware`
4. Create two routes based on the service from each port
5. Verify metrics are returning from the route/metrics endpoint
6. Create an address space and address
7. Verify you get reconciliation metrics for the CRDs like so:
```
controller_runtime_reconcile_total{controller="address-space-controller-controller",result="requeue_after"} 50
controller_runtime_reconcile_total{controller="authenticationservice-controller",result="success"} 1
controller_runtime_reconcile_total{controller="consoleservice-controller",result="requeue_after"} 29
```
and
```
# TYPE controller_runtime_reconcile_time_seconds histogram
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.005"} 46
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.01"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.025"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.05"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.1"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.25"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="0.5"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="1"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="2.5"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="5"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="10"} 50
controller_runtime_reconcile_time_seconds_bucket{controller="address-space-controller-controller",le="+Inf"} 50
controller_runtime_reconcile_time_seconds_sum{controller="address-space-controller-controller"} 0.05244756100000001
controller_runtime_reconcile_time_seconds_count{controller="address-space-controller-controller"} 50
```

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [ ] Write tests and make sure they pass
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging

cc @david-martin
